File: c/handout-01/main.c
#include <sched.h>
#include <stdbool.h>
#include <stdio.h>

#include "eval.h"
#include "parser.h"

static const int variables[255] = {
    ['a'] = 5,
    ['b'] = 7,
    ['c'] = 2,
    ['d'] = 4,
};

void do_eval(const char expr[]) {
  Node* node = node_parse(expr);
  if (node == NULL) {
    printf("\tError: invalid expression given\n");
    return;
  }

  int value = eval(node, variables);
  printf("\tValue = %d\n", value);

  node_free(node);
}

int main() {
  char buf[256];

  while (true) {
    printf("Enter a postfix expression with $ at the end: ");
    scanf("%254s", buf);

    do_eval(buf);

    printf("Continue? (Y/n): ");
    scanf("%254s", buf);

    if (buf[0] == 'n') {
      break;
    }
  }
}

File: py/handout-01/main.py
#-------------------------------------------------------------------------------
#  Group names: Ethan Safai
#  Assignment : No.1
#  Due date   : 02/02/23
# Purpose: this program reads an expression in postfix form, evaluates the 
# expression, and displays its value.
#-------------------------------------------------------------------------------
from postfixparser import PostfixParser

def main():
  """Receive, evaluate, and display the result of postfix expressions from the
  user until they wish to exit the program"""

  parser = PostfixParser()

  while True:
    # read and evaluate user input
    expr = input('Enter a postfix expression with $ at the end: ')
    if not expr.endswith('$') or len(expr) < 2:
      raise Exception('Invalid input')

    result = parser.eval_postfix_expr(expr[:len(expr)-1])
    print(f'Value = {result}')

    # check if the user would like to continue
    if input('CONTINUE(y/n)? ') != 'y':
      break

if __name__ == "__main__":
  main()
